subsys/bluetooth/mesh/proxy.c

```diff
@@ -951,14 +951,15 @@ bool bt_mesh_proxy_relay(struct net_buf *buf, uint16_t dst)
 		net_buf_simple_add_mem(&msg, buf->data, buf->len);
 
 		err = proxy_segment_and_send(client->conn, BT_MESH_PROXY_NET_PDU,
-					&msg, buf_send_end, net_buf_ref(buf));
+					&msg, buf_send_end, buf);
 
 		bt_mesh_adv_send_start(0, err, BT_MESH_ADV(buf));
 		if (err) {
 			BT_ERR("Failed to send proxy message (err %d)", err);
 			continue;
 		}
 
+		net_buf_ref(buf);
```

Comment:
From @jhedberg's review of #35702:
> No code should ever call net_buf_ref() without storing the return value to a new pointer. See my other comment, i.e. the refcounting will need some redesign.

To follow this, I believe this should still be done in the argument as before, and it should unref inside `proxy_segment_and_send` if it fails to deliver it to the bluetooth API.

Reply:
I'm wondering if we should consider annotating ref() functions with `__attribute__ ((warn_unused_result))`.

Reply:
Thumbs up from me at least. It'd be good to have these guidelines documented properly too, since we're blocking reviews with these arguments. The network buffer API documentation just mentions that the refcount concept exists, which isn't really enough to figure out what the guidelines are

Reply:
In fact, I have been troubled by https://github.com/zephyrproject-rtos/zephyr/pull/35702#discussion_r648866718

Reply:
> The network buffer API documentation just mentions that the refcount concept exists, which isn't really enough to figure out what the guidelines are

That's fair, however I don't even consider this net_buf-specific, i.e. there should ideally be common guidelines applied to all ref-counted objects to minimize the chance of bugs and maximize code readability.

Reply:
Out of curiosity, I tried building `tests/bluetooth/shell` with the new GCC attribute and got this:
```
[183/280] Building C object zephyr/subsys/bluetooth/controller/CMakeFiles/subsys__bluetooth__controller.dir/hci/hci_driver.c.obj
/Users/jhedberg/src/zephyr/zephyr/subsys/bluetooth/controller/hci/hci_driver.c: In function 'recv_thread':
/Users/jhedberg/src/zephyr/zephyr/subsys/bluetooth/controller/hci/hci_driver.c:522:4: warning: ignoring return value of 'net_buf_ref', declared with attribute warn_unused_result [-Wunused-result]
    net_buf_ref(frag);
    ^~~~~~~~~~~~~~~~~
[229/280] Building C object zephyr/subsys/bluetooth/host/CMakeFiles/subsys__bluetooth__host.dir/hci_core.c.obj
/Users/jhedberg/src/zephyr/zephyr/subsys/bluetooth/host/hci_core.c: In function 'bt_hci_cmd_send_sync':
/Users/jhedberg/src/zephyr/zephyr/subsys/bluetooth/host/hci_core.c:302:2: warning: ignoring return value of 'net_buf_ref', declared with attribute warn_unused_result [-Wunused-result]
  net_buf_ref(buf);
  ^~~~~~~~~~~~~~~~
```

Reply:
And out of those two at least the second one is quite naturally fixable with this:
```
--- a/subsys/bluetooth/host/hci_core.c
+++ b/subsys/bluetooth/host/hci_core.c
@@ -298,10 +298,7 @@ int bt_hci_cmd_send_sync(uint16_t opcode, struct net_buf *buf,
        k_sem_init(&sync_sem, 0, 1);
        cmd(buf)->sync = &sync_sem;

-       /* Make sure the buffer stays around until the command completes */
-       net_buf_ref(buf);
-
-       net_buf_put(&bt_dev.cmd_tx_queue, buf);
+       net_buf_put(&bt_dev.cmd_tx_queue, net_buf_ref(buf));

        err = k_sem_take(&sync_sem, HCI_CMD_TIMEOUT);
        BT_ASSERT_MSG(err == 0, "k_sem_take failed with err %d", err);
```

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/36259